File: plugin-core/src/main/java/com/github/catatafishen/ideagentforcopilot/settings/ChatWebServerConfigurable.java
package com.github.catatafishen.ideagentforcopilot.settings;

import com.github.catatafishen.ideagentforcopilot.services.ChatWebServer;
import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.project.Project;
import com.intellij.ui.components.JBCheckBox;
import com.intellij.ui.components.JBLabel;
import com.intellij.util.ui.FormBuilder;
import com.intellij.util.ui.JBUI;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.*;
import java.awt.datatransfer.StringSelection;

/**
 * Settings page for the Chat Web Server.
 * Located under Settings → Tools → AgentBridge → Web Access.
 */
public final class ChatWebServerConfigurable implements Configurable {

    private final Project project;
    private JBCheckBox enabledCheckbox;
    private JSpinner portSpinner;
    private JButton startStopButton;
    private JLabel urlLabel;
    private JPanel mainPanel;

    public ChatWebServerConfigurable(@NotNull Project project) {
        this.project = project;
    }

    @Override
    public @Nls(capitalization = Nls.Capitalization.Title) String getDisplayName() {
        return "Web Access";
    }

    @Override
    public @NotNull JComponent createComponent() {
        ChatWebServerSettings settings = ChatWebServerSettings.getInstance(project);

        enabledCheckbox = new JBCheckBox(
            "Start web server automatically when project opens",
            settings.isEnabled());

        portSpinner = new JSpinner(new SpinnerNumberModel(settings.getPort(), 1024, 65535, 1));

        startStopButton = new JButton(getStartStopLabel());
        startStopButton.addActionListener(e -> toggleServer());

        urlLabel = new JBLabel("");
        urlLabel.setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
        updateUrlLabel();

        JButton copyUrlButton = new JButton("Copy URL");
        copyUrlButton.addActionListener(e -> {
            String url = getServerUrl();
            if (!url.isEmpty()) {
                Toolkit.getDefaultToolkit().getSystemClipboard()
                    .setContents(new StringSelection(url), null);
                String orig = copyUrlButton.getText();
                copyUrlButton.setText("Copied!");
                Timer t = new Timer(2000, ev -> copyUrlButton.setText(orig));
                t.setRepeats(false);
                t.start();
            }
        });

        JPanel urlRow = new JPanel(new FlowLayout(FlowLayout.LEFT, 4, 0));
        urlRow.add(urlLabel);
        urlRow.add(copyUrlButton);

        mainPanel = FormBuilder.createFormBuilder()
            .addComponent(new JBLabel("<html>Serve the chat panel as a local web app accessible from "
                + "any device on the same network (phone, tablet, etc.).<br>"
                + "Supports prompt sending, nudging, quick replies, and PWA push notifications.</html>"))
            .addSeparator()
            .addComponent(enabledCheckbox)
            .addLabeledComponent("Port:", portSpinner)
            .addSeparator()
            .addComponent(startStopButton)
            .addComponent(urlRow)
            .addComponentFillVertically(new JPanel(), 0)
            .getPanel();
        mainPanel.setBorder(JBUI.Borders.empty(8));

        reset();
        return mainPanel;
    }

    @Override
    public boolean isModified() {
        ChatWebServerSettings settings = ChatWebServerSettings.getInstance(project);
        if (enabledCheckbox.isSelected() != settings.isEnabled()) return true;
        return (Integer) portSpinner.getValue() != settings.getPort();
    }

    @Override
    public void apply() {
        ChatWebServerSettings settings = ChatWebServerSettings.getInstance(project);
        settings.setEnabled(enabledCheckbox.isSelected());
        settings.setPort((Integer) portSpinner.getValue());
    }

    @Override
    public void reset() {
        ChatWebServerSettings settings = ChatWebServerSettings.getInstance(project);
        enabledCheckbox.setSelected(settings.isEnabled());
        portSpinner.setValue(settings.getPort());
        updateUrlLabel();
        if (startStopButton != null) startStopButton.setText(getStartStopLabel());
    }

    @Override
    public void disposeUIResources() {
        mainPanel = null;
        enabledCheckbox = null;
        portSpinner = null;
        startStopButton = null;
        urlLabel = null;
    }

    private void toggleServer() {
        ChatWebServer ws = ChatWebServer.getInstance(project);
        if (ws.isRunning()) {
            ws.stop();
            refresh();
        } else {
            apply(); // save port before starting
            startStopButton.setEnabled(false);
            startStopButton.setText("Starting…");
            new Thread(() -> {
                try {
                    ws.start();
                } catch (Exception e) {
                    SwingUtilities.invokeLater(() -> {
                        JOptionPane.showMessageDialog(mainPanel,
                            "Failed to start web server: " + e.getMessage(),
                            "Chat Web Server Error", JOptionPane.ERROR_MESSAGE);
                        refresh();
                    });
                    return;
                }
                SwingUtilities.invokeLater(this::refresh);
            }, "ChatWebServer-start").start();
        }
    }

    private void refresh() {
        if (startStopButton == null) return;
        startStopButton.setEnabled(true);
        startStopButton.setText(getStartStopLabel());
        updateUrlLabel();
        if (mainPanel != null) {
            mainPanel.revalidate();
            mainPanel.repaint();
        }
    }

    private String getStartStopLabel() {
        ChatWebServer ws = ChatWebServer.getInstance(project);
        return ws != null && ws.isRunning() ? "Stop Web Server" : "Start Web Server";
    }

    private void updateUrlLabel() {
        if (urlLabel == null) return;
        String url = getServerUrl();
        if (url.isEmpty()) {
            urlLabel.setText("<html><i style='color:gray'>Server not running</i></html>");
        } else {
            urlLabel.setText("<html><a href='" + url + "'>" + url + "</a></html>");
        }
    }

    private String getServerUrl() {
        ChatWebServer ws = ChatWebServer.getInstance(project);
        if (ws == null || !ws.isRunning()) return "";
        try {
            String host = java.net.InetAddress.getLocalHost().getHostAddress();
            return "http://" + host + ":" + ws.getPort();
        } catch (Exception e) {
            return "http://localhost:" + ws.getPort();
        }
    }
}

File: plugin-core/src/main/java/com/github/catatafishen/agentbridge/settings/ChatWebServerConfigurable.kt
package com.github.catatafishen.agentbridge.settings

import com.github.catatafishen.agentbridge.services.ChatWebServer
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.options.BoundConfigurable
import com.intellij.openapi.options.SearchableConfigurable
import com.intellij.openapi.project.Project
import com.intellij.ui.components.JBLabel
import com.intellij.ui.components.JBPanel
import com.intellij.ui.dsl.builder.AlignX
import com.intellij.ui.dsl.builder.bindIntValue
import com.intellij.ui.dsl.builder.bindSelected
import com.intellij.ui.dsl.builder.panel
import java.awt.BorderLayout
import java.awt.Cursor
import java.awt.FlowLayout
import java.awt.Toolkit
import java.awt.datatransfer.StringSelection
import javax.swing.JButton
import javax.swing.JOptionPane
import javax.swing.Timer

class ChatWebServerConfigurable(private val project: Project) :
    BoundConfigurable("Web Access"),
    SearchableConfigurable {

    override fun getId(): String = "com.github.catatafishen.agentbridge.web-access"

    private val settings get() = ChatWebServerSettings.getInstance(project)
    private val urlLabel = JBLabel("").apply { cursor = Cursor.getPredefinedCursor(Cursor.HAND_CURSOR) }
    private val appQrPanel = QrCodePanel()
    private val certQrPanel = QrCodePanel()
    private val certQrRow = buildQrColumn(certQrPanel, "Install CA cert")
    private val appQrRow = buildQrColumn(appQrPanel, "Open on phone")
    private var startStopButton: JButton? = null
    private val copyUrlButton = JButton("Copy URL").also { btn ->
        btn.addActionListener {
            val url = getServerUrl()
            if (url.isNotEmpty()) copyToClipboard(url, btn)
        }
    }

    override fun createPanel() = panel {
        row {
            comment(
                "Serve the chat panel as a local web app accessible from any device on the same network " +
                    "(phone, tablet, etc.). Supports prompt sending, nudging, quick replies, and PWA push notifications."
            )
        }
        separator()
        row {
            checkBox("Start web server automatically when project opens")
                .bindSelected({ settings.isEnabled }, { settings.isEnabled = it })
        }
        row("Port:") {
            spinner(1024..65535, 1)
                .bindIntValue({ settings.port }, { settings.port = it })
        }
        row {
            checkBox("Static port (fail if busy instead of auto-allocating)")
                .comment(
                    "When enabled, the server will not try alternative ports if the configured " +
                        "port is already in use — it will fail with an error instead."
                )
                .bindSelected({ settings.isStaticPort }, { settings.isStaticPort = it })
        }
        row {
            checkBox("Use HTTPS (generates self-signed CA cert for device trust)")
                .bindSelected({ settings.isHttpsEnabled }, { settings.isHttpsEnabled = it })
        }
        separator()
        row {
            startStopButton = button(getStartStopLabel()) { toggleServer() }.component
        }
        row("URL:") {
            cell(urlLabel).align(AlignX.FILL)
            cell(copyUrlButton)
        }
        row {
            cell(appQrRow)
            cell(certQrRow)
        }
        onReset { updateUrlAndQr(); startStopButton?.text = getStartStopLabel() }
    }

    private fun buildQrColumn(qr: QrCodePanel, label: String): JBPanel<*> {
        val col = JBPanel<JBPanel<*>>(BorderLayout())
        val header = JBPanel<JBPanel<*>>(FlowLayout(FlowLayout.LEFT, 0, 0))
        header.add(JBLabel(label))
        col.add(header, BorderLayout.NORTH)
        col.add(qr, BorderLayout.CENTER)
        return col
    }

    private fun toggleServer() {
        val ws = ChatWebServer.getInstance(project)
        if (ws.isRunning) {
            ws.stop(); refresh()
        } else {
            apply()
            startStopButton?.let { it.isEnabled = false; it.text = "Starting…" }
            Thread({
                try {
                    ws.start()
                } catch (e: Exception) {
                    ApplicationManager.getApplication().invokeLater {
                        JOptionPane.showMessageDialog(
                            null,
                            "Failed to start web server: ${e.message}",
                            "Chat Web Server Error", JOptionPane.ERROR_MESSAGE
                        )
                        refresh()
                    }
                    return@Thread
                }
                ApplicationManager.getApplication().invokeLater(::refresh)
            }, "ChatWebServer-start").start()
        }
    }

    private fun refresh() {
        startStopButton?.let { it.isEnabled = true; it.text = getStartStopLabel() }
        updateUrlAndQr()
    }

    private fun getStartStopLabel(): String {
        val ws = ChatWebServer.getInstance(project)
        return if (ws.isRunning) "Stop Web Server" else "Start Web Server"
    }

    private fun updateUrlAndQr() {
        val ws = ChatWebServer.getInstance(project)
        val url = getServerUrl()
        if (url.isEmpty()) {
            urlLabel.text = "<html><i style='color:gray'>Not running</i></html>"
            appQrPanel.setUrl(null); certQrPanel.setUrl(null)
        } else {
            urlLabel.text = "<html><a href='$url'>$url</a></html>"
            appQrPanel.setUrl(url)
            if (isRunningHttps()) {
                val host = ChatWebServer.getLanIp() ?: "localhost"
                certQrPanel.setUrl("http://$host:${ws.httpCertPort}/cert.crt")
            } else {
                certQrPanel.setUrl(null)
            }
        }
        certQrRow.isVisible = isRunningHttps()
    }

    private fun getServerUrl(): String {
        val ws = ChatWebServer.getInstance(project)
        if (!ws.isRunning) return ""
        val protocol = if (ws.isHttps) "https" else "http"
        val host = ChatWebServer.getLanIp() ?: "localhost"
        return "$protocol://$host:${ws.port}"
    }

    private fun isRunningHttps(): Boolean {
        val ws = ChatWebServer.getInstance(project)
        return ws.isRunning && ws.isHttps
    }

    private fun copyToClipboard(text: String, feedbackButton: JButton) {
        Toolkit.getDefaultToolkit().systemClipboard.setContents(StringSelection(text), null)
        val orig = feedbackButton.text
        feedbackButton.text = "Copied!"
        Timer(2000) { feedbackButton.text = orig }.apply { isRepeats = false }.start()
    }
}
